waspc/src/Wasp/AI/CodeAgent.hs

```diff
@@ -35,33 +35,33 @@ import qualified Wasp.Util.IO.Retry as R
 import Wasp.Util.Network.HTTP (catchRetryableHttpException)
 import qualified Wasp.Util.Network.HTTP as Utils.HTTP
 
-newtype CodeAgent a = CodeAgent {_unCodeAgent :: ReaderT CodeAgentConfig (StateT CodeAgentState IO) a}
-  deriving (Monad, Applicative, Functor, MonadIO, MonadReader CodeAgentConfig, MonadState CodeAgentState)
+newtype CodeAgent logMsg a = CodeAgent {_unCodeAgent :: ReaderT (CodeAgentConfig logMsg) (StateT CodeAgentState IO) a}
```

Comment:
I don't understand the `logMsg` parameter, it has no constraints on it?

Reply:
So `CodeAgent` here is a standalone module, it could be in theory extracted in to a library -> it doesn't know anything about how we use it in Mage.

I could have made it directly use LogMsg data type which comes with it, but I don't really need to be that specific, instead I use anything that implements IsString -> that is enough for CogeAgent to know what to do with it. This in theory allows using CodeAgent with one type of log messages in one place, and with other kind of log message in some other use case.

Now, why there are no restrictions on the data type -> that is a bit non-instinctive in Haskell, but actually it is much better to put type constraints on functions that use the data type, not on data type itself. SO you will see I have `IsString` constraint for `logMsg` on some functions in this file. This has some benefits to it and is how it is done. If I am correct, putting constraints on the data type actually won't even have the wanted effect. We can discuss this more if you are interested into this topic.